provider/kubernetes/client.go

```diff
@@ -186,109 +162,51 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
 	return eventCh, nil
 }
 
-// WatchIngresses sets up a watch on Ingress objects and returns a corresponding shared informer.
-func (c *clientImpl) WatchIngresses(namespace string, labelSelector labels.Selector, watchCh chan<- interface{}) cache.SharedInformer {
-	listOptions := metav1.ListOptions{
-		LabelSelector: labelSelector.String(),
-		FieldSelector: fields.Everything().String(),
-	}
-	informer := loadInformer(&cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			return c.clientset.ExtensionsV1beta1().Ingresses(namespace).List(listOptions)
-		},
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			return c.clientset.ExtensionsV1beta1().Ingresses(namespace).Watch(listOptions)
-		},
-	}, &extensionsv1beta1.Ingress{}, watchCh)
-	c.ingStores = append(c.ingStores, informer.GetStore())
-	return informer
-}
-
-// WatchObjects sets up a watch on objects and returns a corresponding shared informer.
-func (c *clientImpl) WatchObjects(namespace, kind string, object runtime.Object, storeMap map[string]cache.Store, watchCh chan<- interface{}) cache.SharedInformer {
-	listWatch := cache.NewListWatchFromClient(
-		c.clientset.CoreV1().RESTClient(),
-		kind,
-		namespace,
-		fields.Everything())
-
-	informer := loadInformer(listWatch, object, watchCh)
-	storeMap[namespace] = informer.GetStore()
-	return informer
-}
-
-func loadInformer(listWatch cache.ListerWatcher, object runtime.Object, watchCh chan<- interface{}) cache.SharedInformer {
-	informer := cache.NewSharedInformer(
-		listWatch,
-		object,
-		resyncPeriod,
-	)
-
-	informer.AddEventHandler(newResourceEventHandler(watchCh))
-	return informer
-}
-
 // GetIngresses returns all Ingresses for observed namespaces in the cluster.
 func (c *clientImpl) GetIngresses() []*extensionsv1beta1.Ingress {
 	var result []*extensionsv1beta1.Ingress
-
-	for _, store := range c.ingStores {
-		for _, obj := range store.List() {
-			ing := obj.(*extensionsv1beta1.Ingress)
+	for ns, factory := range c.factories {
+		ings, err := factory.Extensions().V1beta1().Ingresses().Lister().List(labels.Everything())
+		if err != nil {
+			log.Errorf("fail to list ingresses in namespace %s: %s", ns, err)
+		}
+		for _, ing := range ings {
 			result = append(result, ing)
 		}
 	}
-
 	return result
 }
 
 // GetService returns the named service from the given namespace.
 func (c *clientImpl) GetService(namespace, name string) (*corev1.Service, bool, error) {
 	var service *corev1.Service
-	item, exists, err := c.svcStores[c.lookupNamespace(namespace)].GetByKey(namespace + "/" + name)
+	item, exists, err := c.factories[namespace].Core().V1().Services().Informer().GetStore().GetByKey(namespace + "/" + name)
```

Comment:
Ideally this line could be `svc, err := c.factories[namespace].Core().V1().Services().Lister().Services(namespace).Get(name)` which use `Lister` to `Get` resources from indexer inside the informer. But this will need to change the return types of `Client` interface to `GetService(namespace, name string) (*corev1.Service, error)` and lots of changes in test facilities will change too.

Reply:
Note that they are actually the same. But eliminating `namespace + "/" + name` will help make the code clearer.